### PR TITLE
feat: custom compiler

### LINKED ISSUE #109: Icon Component overhead
I will try explain using Svelte as an example, but it should be the same.

Currently import an an icon, unplugin-icons will generate an component on the fly, like this 

```js
import IconComponent from '~icons/collection/name
```
 to this IconComponent.svelte
```svelte
<svg viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg" {...$$props}>{@html `<path d="xxx" />`}</svg>
```
when bundling, it will compile to this for every component

```js
function nt(t) {
	let e,
		n =
			'<path d="xxx"/>',
		r = [
			{ viewBox: '0 0 24 24' },
			{ fill: 'none' },
			{ xmlns: 'http://www.w3.org/2000/svg' },
			t[0],
		],
		s = {};
	for (let o = 0; o < r.length; o += 1) s = d(s, r[o]);
	return {
		c() {
			(e = V('svg')), _(e, s);
		},
		m(o, c) {
			A(o, e, c), (e.innerHTML = n);
		},
		p(o, [c]) {
			_(
				e,
				(s = P(r, [
					{ viewBox: '0 0 24 24' },
					{ fill: 'none' },
					{ xmlns: 'http://www.w3.org/2000/svg' },
					c & 1 && o[0],
				]))
			);
		},
		i: f,
		o: f,
		d(o) {
			o && k(e);
		},
	};
}
```
this is what I use for my projects to minify overhead and without losing reactivity (not 100% sure I'm right though)

A shared Icon component

```svelte
<script>
    export let icon;
</script>

<svg xmlns="http://www.w3.org/2000/svg" {...icon.props} {...$$restProps}>
    {@html icon.path}
</svg>
```

and compile svg string to obj like this
```js
export default {
    path: `<path d="xxx" />`,
    props: {
         fill: 'none',
         viewBox: '0 0 24 24',
    }
}
```
use it like this
```svelte
<Icon icon={checked? icon1: icon2} height={size} width={size} on:click={onClick}/>
```

I think this way will save some code size, and keep reactivity, ssr should be ok too.

I know this way it may go against the main purpose of this project, use icon as Component, but maybe it can be an option

- add object Compiler, svg to icon object
- auto generate shared Icon component `~icons/component/svelte`
- maybe auto importing can work too, not sure

### LINKED ISSUE #111: feat: custom icon compiler
As described in https://github.com/antfu/unplugin-icons/issues/109  there are situations where Application built in compiler based frameworks want to utilize custom icon components.

Thess custom compilers could be provided inline in the plugin config or via 3rd party packages.
To enable using them the `compiler: string` option could be changed to allow  passing a compiler function directly.

eg

```js
Icons({
  compiler: svg => { const compiledSvg='do stuff'; return compiledSvg;}
})
```

```js
import customCompiler from 'some-package'
Icons({
  compiler: customCompiler()
})
```

### PR DESCRIPTION
**Pull request by userquin**
closes #109 
closes #111

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -14,17 +14,26 @@ const unplugin = createUnplugin<Options>((options = {}) => {
...
-            return `${res}.jsx`
-          case 'svelte':
-            return `${res}.svelte`
-          case 'solid':
-            return `${res}.tsx`
-          case 'marko':
-            return `${res}.marko`
-          default:
-            return res
+        if (options.compiler) {
```
Comment (antfu) [inline src/index.ts]:
```ts
if (options.compiler && typeof options.compiler !== 'string') {
  // custom
}
else {
  switch(options.compiler || '') {

  }
}
```

Comment (userquin) [reply src/index.ts]: see changes

File: src/core/loader.ts [hunk 2]
```diff
@@ -91,7 +91,10 @@ export async function generateComponent({ collection, icon }: ResolvedIconPath,
   if (defaultStyle)
     svg = svg.replace('<svg ', `<svg style="${defaultStyle}" `)
 
-  const compiler = compilers[options.compiler]
+  const compiler = typeof options.compiler === 'string'
+    ? compilers[options.compiler]
+    : (await options.compiler.compiler)
```
Comment (dominikg) [inline src/core/loader.ts:96]: should this be `(await options.compiler).compiler` ?

Comment (userquin) [reply src/core/loader.ts:96]: good catch

Comment (userquin) [reply src/core/loader.ts]: it was ok, the first compiler is from options and the second one is the awaitable compiler on CustomCompiler: return an Awaitable<string>....

Comment (userquin) [reply src/core/loader.ts]: I need to revert it

**Review by dominikg — APPROVED** [review]

**Comment by userquin** [comment]
@antfu the usage is weird, maybe we can chnage the signature for `compiler` on `CustomCompiler` to:
```ts
export type CustomCompiler = {
  compiler: (() => Awaitable<Compiler>) | Compiler
  extension?: string
}
```

**Review by antfu — APPROVED** [review]

**Pull request merged**